src/gs-shell.c

```diff
@@ -853,6 +877,125 @@ gs_shell_show_search_result (GsShell *shell, const gchar *id, const gchar *searc
 	gs_shell_change_mode (shell, GS_SHELL_MODE_SEARCH, search, TRUE);
 }
 
+void
+gs_shell_side_filter_set_visible (GsShell *shell, gboolean visible)
+{
+	GsShellPrivate *priv = gs_shell_get_instance_private (shell);
+	gtk_widget_set_visible (priv->side_filter_scrolled_window, visible);
+}
+
+static void
+side_filter_row_selected (GtkListBox *side_filter,
+			  GsSideFilterRow *row,
+			  gpointer data)
+{
+	GsShell *shell = GS_SHELL (data);
+	GsShellMode mode = gs_side_filter_row_get_mode (row);
+	GsCategory *category = NULL;
+
+	switch (mode) {
+	case GS_SHELL_MODE_OVERVIEW:
+		gs_shell_set_mode (shell, GS_SHELL_MODE_OVERVIEW);
+		break;
+	case GS_SHELL_MODE_CATEGORY:
+		category = gs_side_filter_row_get_category (row);
+		gs_shell_show_category (shell, category);
+		break;
+	default:
+		g_assert_not_reached ();
+	}
+}
+
+static void
+side_filter_select_by_mode (GsShell *shell, GsShellMode mode)
+{
+	g_autoptr(GList) rows = NULL;
+	GList *l;
+	GsSideFilterRow *row;
+	GsShellPrivate *priv = gs_shell_get_instance_private (shell);
+
+	row = GS_SIDE_FILTER_ROW (gtk_list_box_get_selected_row (GTK_LIST_BOX (priv->side_filter)));
+
+	/* if the mode if already selected, we do nothing; this not only
+	 * optimizes things but also prevents us from always selecting the
+	 * first category-mode row when a category view is displayed */
+	if (row && gs_side_filter_row_get_mode (row) == mode) {
+		gs_shell_side_filter_set_visible (shell, TRUE);
+		return;
+	}
+
+	row = NULL;
+
+	/* we don't want it to select a row and switch to the view we're
+	 * already in so disconnect the handler functions */
+	g_signal_handlers_block_by_func (priv->side_filter,
+					 side_filter_row_selected, shell);
+
+	rows = gtk_container_get_children (GTK_CONTAINER (priv->side_filter));
+	for (l = rows; l; l = l->next) {
+		GsSideFilterRow *current = GS_SIDE_FILTER_ROW (l->data);
+		if (gs_side_filter_row_get_mode (current) == mode) {
+			row = current;
+			break;
+		}
+	}
+
+	if (row) {
+		gtk_list_box_select_row (GTK_LIST_BOX (priv->side_filter),
+					 GTK_LIST_BOX_ROW (row));
+	}
+
+	g_signal_handlers_unblock_by_func (priv->side_filter,
+					   side_filter_row_selected, shell);
+
+	gs_shell_side_filter_set_visible (shell, (row != NULL));
+}
+
+GtkWidget *
+gs_shell_side_filter_add_category (GsShell *shell, GsCategory *cat)
+{
+	GtkWidget *row;
+	GsShellPrivate *priv = gs_shell_get_instance_private (shell);
+
+	row = gs_side_filter_row_new (cat);
+	gtk_list_box_insert (GTK_LIST_BOX (priv->side_filter), row, -1);
+
+	return row;
+}
+
+static void
+gs_shell_side_filter_add_mode (GsShell *shell,
```

Comment:
Then I would just return instead of asserting the passed-in mode is the overview one.